R/helper_clean_model_data.R

```diff
@@ -129,3 +153,33 @@ clean_model_data <- function(data, datargs, estimator = "") {
 
   return(ret)
 }
+
+demean_fes <- function(model_data) {
+  nfaclevels <-
+    apply(model_data[["fixed_effects"]], 2, function(fe) length(unique((fe)))-1)
+
+  demeaned <- demeanMat(
+    Y = as.matrix(model_data[["outcome"]]),
+    X = model_data[["design_matrix"]],
+    Z = model_data[["instrument_matrix"]],
+    fes = model_data[["fixed_effects"]],
+    weights = if (is.numeric(model_data[["weights"]])) model_data[["weights"]] else rep(1, nrow(model_data[["design_matrix"]])),
+    has_int = attr(model_data$terms, "intercept"),
+    eps = 1e-8
+  )
+
+  # save names
+  dimnames(demeaned[["newY"]]) <- dimnames(model_data[["outcome"]])
+  new_names <- dimnames(model_data[["design_matrix"]])
+  new_names[[2]] <- new_names[[2]][new_names[[2]] != "(Intercept)"]
+  dimnames(demeaned[["newX"]]) <- new_names
+
+  model_data[["outcome"]] <- demeaned[["newY"]]
+  model_data[["design_matrix"]] <- demeaned[["newX"]]
```

Comment:
Ideally, match up the names on demeaned with model_data - eg demeaned$outcome instead of demeaned$newY etc.